File: src/main/generic/SDL_sysmain_callbacks.c
/*
  Simple DirectMedia Layer
  Copyright (C) 1997-2023 Sam Lantinga <slouken@libsdl.org>

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.
*/

#include "SDL_internal.h"
#include "../SDL_main_callbacks.h"
#include "../../video/SDL_sysvideo.h"

#ifndef __IOS__

static int callback_rate_increment = 0;

static void SDLCALL MainCallbackRateHintChanged(void *userdata, const char *name, const char *oldValue, const char *newValue)
{
    const int callback_rate = newValue ? SDL_atoi(newValue) : 60;
    if (callback_rate > 0) {
        callback_rate_increment = ((Uint64) 1000000000) / ((Uint64) callback_rate);
    } else {
        callback_rate_increment = 0;
    }
}

int SDL_EnterAppMainCallbacks(int argc, char* argv[], SDL_AppInit_func appinit, SDL_AppIterate_func appiter, SDL_AppEvent_func appevent, SDL_AppQuit_func appquit)
{
    int rc = SDL_InitMainCallbacks(argc, argv, appinit, appiter, appevent, appquit);
    if (rc == 0) {
        SDL_AddHintCallback(SDL_HINT_MAIN_CALLBACK_RATE, MainCallbackRateHintChanged, NULL);

        Uint64 next_iteration = callback_rate_increment ? (SDL_GetTicksNS() + callback_rate_increment) : 0;

        while ((rc = SDL_IterateMainCallbacks(SDL_TRUE)) == 0) {
            // !!! FIXME: this can be made more complicated if we decide to
            // !!! FIXME: optionally hand off callback responsibility to the
            // !!! FIXME: video subsystem (for example, if Wayland has a
            // !!! FIXME: protocol to drive an animation loop, maybe we hand
            // !!! FIXME: off to them here if/when the video subsystem becomes
            // !!! FIXME: initialized).

            // !!! FIXME: maybe respect this hint even if there _is_ a window.
            // if there's no window, try to run at about 60fps (or whatever rate
            //  the hint requested). This makes this not eat all the CPU in
            //  simple things like loopwave. If there's a window, we run as fast
            //  as possible, which means we'll clamp to vsync in common cases,
            //  and won't be restrained to vsync if the app is doing a benchmark
            //  or doesn't want to be, based on how they've set up that window.
            if ((callback_rate_increment == 0) || SDL_HasWindows()) {
                next_iteration = 0; // just clear the timer and run at the pace the video subsystem allows.
            } else {
                const Uint64 now = SDL_GetTicksNS();
                if (next_iteration > now) { // Running faster than the limit, sleep a little.
                    SDL_DelayNS(next_iteration - now);
                } else {
                    next_iteration = now; // running behind (or just lost the window)...reset the timer.
                }
                next_iteration += callback_rate_increment;
            }
        }

        SDL_DelHintCallback(SDL_HINT_MAIN_CALLBACK_RATE, MainCallbackRateHintChanged, NULL);
    }
    SDL_QuitMainCallbacks();

    return (rc < 0) ? 1 : 0;
}

#endif // !__IOS__

File: src/main/ios/SDL_sysmain_callbacks.m
/*
  Simple DirectMedia Layer
  Copyright (C) 1997-2024 Sam Lantinga <slouken@libsdl.org>

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.
*/

#include "SDL_internal.h"
#include "../SDL_main_callbacks.h"

#ifdef SDL_PLATFORM_IOS

#import <UIKit/UIKit.h>

@interface SDLIosMainCallbacksDisplayLink : NSObject
@property(nonatomic, retain) CADisplayLink *displayLink;
- (void)appIteration:(CADisplayLink *)sender;
- (instancetype)init:(SDL_AppIterate_func)_appiter quitfunc:(SDL_AppQuit_func)_appquit;
@end

static SDLIosMainCallbacksDisplayLink *globalDisplayLink;

@implementation SDLIosMainCallbacksDisplayLink

- (instancetype)init:(SDL_AppIterate_func)_appiter quitfunc:(SDL_AppQuit_func)_appquit;
{
    if ((self = [super init])) {
        self.displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(appIteration:)];
        [self.displayLink addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];
    }
    return self;
}

- (void)appIteration:(CADisplayLink *)sender
{
    const int rc = SDL_IterateMainCallbacks(SDL_TRUE);
    if (rc != 0) {
        [self.displayLink invalidate];
        self.displayLink = nil;
        globalDisplayLink = nil;
        SDL_QuitMainCallbacks();
        exit((rc < 0) ? 1 : 0);
    }
}
@end

// SDL_RunApp will land in UIApplicationMain, which calls SDL_main from postFinishLaunch, which calls this.
// When we return from here, we're living in the RunLoop, and a CADisplayLink is firing regularly for us.
int SDL_EnterAppMainCallbacks(int argc, char* argv[], SDL_AppInit_func appinit, SDL_AppIterate_func appiter, SDL_AppEvent_func appevent, SDL_AppQuit_func appquit)
{
    const int rc = SDL_InitMainCallbacks(argc, argv, appinit, appiter, appevent, appquit);
    if (rc == 0) {
        globalDisplayLink = [[SDLIosMainCallbacksDisplayLink alloc] init:appiter quitfunc:appquit];
        if (globalDisplayLink != nil) {
            return 0;  // this will fall all the way out of SDL_main, where UIApplicationMain will keep running the RunLoop.
        }
    }

    // appinit requested quit, just bounce out now.
    SDL_QuitMainCallbacks();
    exit((rc < 0) ? 1 : 0);

    return 1;  // just in case.
}

#endif


